src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -3283,6 +3283,174 @@ def groupby_apply(
             )
         )
 
+    def _fill_null_values_for_groupby_first_last(
+        self, method: str, by_list: list[str]
+    ) -> dict[str, ColumnOrName]:
+        """
+        Fill null values in each column using method within each group.
+
+        Args:
+            method: "bfill" or "ffill"
+                The method to use to fill null values.
+            by_list: list[str]
+                The list of columns to partition by during the fillna.
+
+        Returns:
+            dict: A mapping between column name and the Snowpark Column object with
+                replaced null values.
+        """
+        method = FillNAMethod.get_enum_for_string_method(method)
+        method_is_ffill = method is FillNAMethod.FFILL_METHOD
+        if method_is_ffill:
+            func = last_value
+            window_start = Window.UNBOUNDED_PRECEDING
+            window_end = Window.CURRENT_ROW
+        else:
+            func = first_value
+            window_start = Window.CURRENT_ROW
+            window_end = Window.UNBOUNDED_FOLLOWING
+
+        #
+        return {
+            snowflake_quoted_id: coalesce(
+                snowflake_quoted_id,
+                func(snowflake_quoted_id, ignore_nulls=True).over(
+                    Window.partition_by(by_list)
+                    .order_by(
+                        self._modin_frame.row_position_snowflake_quoted_identifier
+                    )
+                    .rows_between(window_start, window_end)
+                ),
+            )
+            for snowflake_quoted_id in self._modin_frame.data_column_snowflake_quoted_identifiers
+        }
+
+    def groupby_first(
+        self,
+        by: Any,
+        axis: int,
+        groupby_kwargs: dict[str, Any],
+        agg_args: tuple[Any],
+        agg_kwargs: dict[str, Any],
+        drop: bool = False,
+        **kwargs: dict[str, Any],
+    ) -> "SnowflakeQueryCompiler":
+        """
+        Get the first non-null value for each group.
+
+        Args:
+            by: mapping, series, callable, label, pd.Grouper, BaseQueryCompiler, list of such.
+                Use this to determine the groups.
+            axis: 0 (index) or 1 (columns).
+            groupby_kwargs: dict
+                keyword arguments passed for the groupby.
+            agg_args: tuple
+                The aggregation args, unused in `groupby_size`.
+            agg_kwargs: dict
+                The aggregation keyword args, unused in `groupby_size`.
+            drop: bool
+                Drop the `by` column, unused in `groupby_size`.
+
+        Returns:
+            SnowflakeQueryCompiler: The result of groupby_first()
+        """
+        level = groupby_kwargs.get("level", None)
```

Comment:
Can we create one more helper method that encapsulates the whole function, with a parameter input for `first` and `last`? Seems like there is a lot of repeated code here.